### PR TITLE
[Quantization] Validate checkpointing works in mix mode quant

### PR DESCRIPTION
**Pull request by kimishpatel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111923

Summary:
Adding a test to ensure that when training a model with qat and applying
ptq later is still compatible with checkpointing

Test Plan:
Test added

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D50602861](https://our.internmc.facebook.com/intern/diff/D50602861)

### CODE REVIEW

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50602861).

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50602861).

File: test/quantization/pt2e/test_quantize_pt2e_qat.py [hunk 1]
```diff
@@ -732,3 +732,78 @@ def test_mixing_qat_ptq(self):
...
+            after_prepare_result_pt2e = model(*example_inputs)
+            target = torch.rand_like(after_prepare_result_pt2e)
+            output = loss(after_prepare_result_pt2e, target)
+            output.backward()
+            optimizer.step()
+
+        state_dict_before = copy.deepcopy(model.state_dict())
+        torch.save(model.state_dict(), buffer)
+        buffer.seek(0)
+        model.load_state_dict(torch.load(buffer))
```
Comment (jerryzh168) [inline test/quantization/pt2e/test_quantize_pt2e_qat.py:758]: nit: to simulate the real use case, this model should be initialized from the beginning I think

File: test/quantization/pt2e/test_quantize_pt2e_qat.py [hunk 2]
```diff
@@ -732,3 +732,78 @@ def test_mixing_qat_ptq(self):
...
+            example_inputs,
+        )
+
+        quantizer = XNNPACKQuantizer()
+        quantizer.set_module_type(torch.nn.Linear, None)
+        quantization_config = get_symmetric_quantization_config()
+        quantizer.set_global(quantization_config)
+        model_pt2e = prepare_pt2e(model_pt2e, quantizer)
+        after_prepare_result_pt2e = model_pt2e(*example_inputs)
+        model_pt2e = convert_pt2e(model_pt2e)
```
Comment (jerryzh168) [inline test/quantization/pt2e/test_quantize_pt2e_qat.py:788]: fold_quantize=True, maybe we should prioritize the BE work to remove this flag?

Comment (kimishpatel) [reply test/quantization/pt2e/test_quantize_pt2e_qat.py:788]: Thanks. Keep forgetting

File: test/quantization/pt2e/test_quantize_pt2e_qat.py (hunk 1 above)
Comment (kimishpatel) [reply test/quantization/pt2e/test_quantize_pt2e_qat.py:758]: what do you mean?

Comment (jerryzh168) [reply test/quantization/pt2e/test_quantize_pt2e_qat.py:758]: oh I mean we should repeat L741-744 here before loading the state_dict to simulate the use case that people do checkpointing, and resume training a bit later

Comment (kimishpatel) [reply test/quantization/pt2e/test_quantize_pt2e_qat.py:761]: Thanks for pointing this out. It does require running through prepare_qat step again.

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50602861).

File: test/quantization/pt2e/test_quantize_pt2e_qat.py [hunk 3]
```diff
@@ -720,7 +720,88 @@ def test_mixing_qat_ptq(self):
...
+        loss = torch.nn.MSELoss()
+        optimizer = torch.optim.SGD(model.parameters(), lr=0.001, momentum=0.9)
+        for i in range(2):
+            after_prepare_result_pt2e = model(*example_inputs)
+            target = torch.rand_like(after_prepare_result_pt2e)
+            output = loss(after_prepare_result_pt2e, target)
+            output.backward()
+            optimizer.step()
+
+        state_dict_before = copy.deepcopy(model.state_dict())
```
Comment (jerryzh168) [inline test/quantization/pt2e/test_quantize_pt2e_qat.py:755]: nit: could you also add some comments to specify the context for each section of code? like QAT, checkpoint, load checkpoint, PTQ etc.

**Review by jerryzh168 — APPROVED** [review]

**Pull request closed without merging**